components/servo_tracing/src/lib.rs

```diff
@@ -0,0 +1,319 @@
+        field_args,
+        level,
+        name,
+    } = parse2(attr)?;
+    let input_fn: ItemFn = parse2(item)?;
+
+    // setup invariants to make reconstruction easier
+    let field_args = field_args.unwrap_or_else(|| parse_quote! { fields(servo_profiling = true) });
+    let field_args = inject_servo_profiling(field_args)?;
+    // invariant, field_args is `fields(...)` and `servo_profiling = true` is injected unless explicitly
+    // set otherwise
+    let level = level.unwrap_or(parse2(quote! {"trace"}).unwrap());
+    passthrough_args.push(parse2(quote! {level = #level}).unwrap());
+    // invariant, passthrough args has >= 1 items
+    let passthrough_args = quote! {
+        #(#passthrough_args),*,
+    };
+    // invariant, passthrough_args has a comma
+    let name = match name {
+        Some(name) => quote! { #name, },
+        None => quote! {},
+    };
+    // invariant, name has a comma if needed
+
+    // reconstruct
+    let fn_attrs: &Vec<Attribute> = &input_fn.attrs;
+    let fn_vis = &input_fn.vis;
+    let fn_sig = &input_fn.sig;
+    let fn_block = &input_fn.block;
+
+    let output = quote! {
+        #[cfg_attr(
+            feature = "tracing",
+            tracing::instrument(
+                #name
+                #passthrough_args
+                #field_args
+            )
+        )]
+        #(#fn_attrs)*
+        #fn_vis #fn_sig #fn_block
+    };
+
+    Ok(output)
+}
+
+#[proc_macro_attribute]
+/// Instruments a function with some sane defaults by automatically:
+///  - setting the attribute behind the "tracing" flag
+///  - adding `servo_profiling = true` in the `tracing::instrument(fields(...))` argument.
+///  - setting `level = "trace"` if it is not given.
+/// We need to be able to set the following
+/// ```
+/// #[cfg_attr(
+///         feature = "tracing",
+///         tracing::instrument(
+///             name = "MyCustomName",
+///             skip_all,
+///             fields(servo_profiling = true),
+///             level = "trace",
+///         )
+///     )]
+/// fn my_fn() { /* .... */ }
+/// ```
+/// from a simpler macro, such as:
+///
+/// ```
+/// #[servo_tracing::instrument(name = "MyCustomName", skip_all)]
+/// fn my_fn() { /* .... */ }
+/// ```
+pub fn instrument(attr: TokenStream, item: TokenStream) -> TokenStream {
+    match instrument_internal(attr.into(), item.into()) {
+        Ok(stream) => stream.into(),
+        Err(err) => err.to_compile_error().into(),
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use crate::{instrument, instrument_internal};
+    use proc_macro2::TokenStream;
+    use quote::{ToTokens, quote};
+    use std::io::Write;
+    use std::process::{Command, Stdio};
+    use syn::{Attribute, ItemFn, parse2, token::Group};
+
+    /// to reduce issues with comparisons, we format the strings; this gives us some invariance wrt
+    /// formatting.
+    fn rustfmt_tokenstream(ts: proc_macro2::TokenStream) -> String {
+        let mut child = Command::new("rustfmt")
```

Comment:
How about using [prettyplease](https://docs.rs/prettyplease/latest/prettyplease/) instead and avoid shelling out.